test/Microsoft.AspNetCore.Razor.Test.Common/Language/Legacy/ClassifiedSpanWriter.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy
+{
+    internal class ClassifiedSpanWriter
+    {
+        private readonly TextWriter _writer;
+
+        public ClassifiedSpanWriter(TextWriter writer)
+        {
+            _writer = writer;
+        }
+
+        public virtual void Visit(SyntaxTreeNode node)
+        {
+            var classifiedSpans = GetClassifiedSpans(node);
+            foreach (var span in classifiedSpans)
+            {
+                VisitClassifiedSpan(span);
+                WriteNewLine();
+            }
+        }
+
+        public virtual void VisitClassifiedSpan(ClassifiedSpanInternal span)
+        {
+            WriteClassifiedSpan(span);
+        }
+
+        protected void WriteClassifiedSpan(ClassifiedSpanInternal span)
+        {
+            Write($"{span.SpanKind} span at {span.Span} (Accepts:{span.AcceptedCharacters})");
+            WriteSeparator();
+            Write($"Parent: {span.BlockKind} block at {span.BlockSpan}");
+        }
+
+        protected void WriteSeparator()
+        {
+            Write(" - ");
+        }
+
+        protected void WriteNewLine()
+        {
+            _writer.WriteLine();
+        }
+
+        protected void Write(object value)
+        {
+            _writer.Write(value);
+        }
+
+        internal static IReadOnlyList<ClassifiedSpanInternal> GetClassifiedSpans(SyntaxTreeNode root)
+        {
+            var spans = Flatten(root);
+
+            var result = new ClassifiedSpanInternal[spans.Count];
+            for (var i = 0; i < spans.Count; i++)
+            {
+                var span = spans[i];
+                result[i] = new ClassifiedSpanInternal(
+                    new SourceSpan(
+                        span.Start.FilePath,
+                        span.Start.AbsoluteIndex,
+                        span.Start.LineIndex,
+                        span.Start.CharacterIndex,
+                        span.Length),
+                    new SourceSpan(
+                        span.Parent.Start.FilePath,
+                        span.Parent.Start.AbsoluteIndex,
+                        span.Parent.Start.LineIndex,
+                        span.Parent.Start.CharacterIndex,
+                        span.Parent.Length),
+                    span.Kind,
+                    span.Parent.Type,
+                    span.EditHandler.AcceptedCharacters);
+            }
+
+            return result;
+        }
+
+        private static List<Span> Flatten(SyntaxTreeNode root)
```

Comment:
Does this duplicate product code? Should we add it to `SyntaxTreeNode`?

Reply:
It does. And I don't see why not make it an extension method. 👍

Reply:
It should just be a normal extension method. 

What do you stand to gain besides #PranavPoints for making it an extension method?

Reply:
In general not a huge fan of having these tooling specific baselines in the language project. In the end this is all to ensure tooling works. It'd be nice for this all to eventually sit in the tooling test projects and just link the appropriate files.

Anyways, rant over, good enough for now.

Reply:
So your suggestion is to move this feature down the syntax tree layer? sounds like a good idea :trollface:

Reply:
Moving these baselines to a tooling test assembly would also mean copying our existing parser tests for which we generate the baselines. I would much prefer they stay and be part of the same tests so it's much easier to run, maintain and find out when something changes.

Reply:
> So your suggestion is to move this feature down the syntax tree layer?

I'm not suggesting moving any feature; just suggesting to move the tests to the appropriate test dll.  As this PR currently stands the baselines are only verifying the copied code from the tooling binaries. 

Basically, what i have beef with is that in this PRs current state the test baseline logic fakes what the `Microsoft.VisualStudio.Editor.Razor.dll (MVSER)` syntax facts service does; it doesn't actually call through into the code that the WTE code calls into. Looking further into the future, with this approach the code paths will further diverge because we're going to need some sort of adapter in the `MVSER.dll` that understands the new syntax tree to ensure that it spits out 100% identical spans prior to the parser re-design.